pkg/controller/backup_sync_controller.go

```diff
@@ -262,6 +272,70 @@ func (c *backupSyncController) run() {
 					log.Debug("Synced pod volume backup into cluster")
 				}
 			}
+
+			if features.IsEnabled(velerov1api.CSIFeatureFlag) {
+				log.Info("Syncing CSI snapshot API objects in backup")
+				snapConts, err := backupStore.GetCSIVolumeSnapshotContents(backupName)
+				// if we fail to get the CSI volumesnapshotcontents from the backup we will
+				// also skip syncing the CSI volumesnapshots because just having the CSI volumesnapshot object
+				// is of no use.
+				if err != nil {
+					log.WithError(errors.WithStack(err)).Error("Error getting CSI volumesnapshotcontents for this backup from backup store")
+					continue
+				}
+
+				log.Infof("Syncing CSI %d volumesnapshotcontents in backup", len(snapConts))
+				for _, snapCont := range snapConts {
+					// TODO: Reset ResourceVersion prior to persisting VolumeSnapshotContents
+					snapCont.ResourceVersion = ""
+					created, err := c.csiSnapshotClient.SnapshotV1beta1().VolumeSnapshotContents().Create(snapCont)
+					switch {
+					case err != nil && kuberrs.IsAlreadyExists(err):
+						log.Debugf("volumesnapshotcontent %s already exists in cluster", snapCont.Name)
+						continue
+					case err != nil && !kuberrs.IsAlreadyExists(err):
+						log.WithError(errors.WithStack(err)).Errorf("Error syncing volumesnapshotcontent %s into cluster", snapCont.Name)
+						continue
+					default:
+						log.Infof("Created CSI volumesnapshotcontent %s", created.Name)
+					}
+				}
+
+				csiVolSnaps, err := backupStore.GetCSIVolumeSnapshots(backupName)
+				if err != nil {
+					log.WithError(errors.WithStack(err)).Error("Error getting CSI volumesnapshots for this backup from backup store")
+					continue
+				}
+
+				log.Infof("Syncing %d CSI volumesnapshots in backup", len(csiVolSnaps))
```

Comment:
Wouldn't we need to do the whole dynamic -> static provisioning change here like we do for the plugins?

Reply:
I don't think that would be necessary, because:
1. With syncing a backup into the cluster, we are merely re-creating the state that was captured while performing the backup and not exactly restoring the backup.
2. Also, the motivation to sync the CSI API objects is to ensure that the storage snapshots are cleaned up on backup deletion/expiry.
3. We will be doing the dynamic -> static provisioning sequence when we restore from the backup.

Reply:
OK - I was under the impression we'd need to change the VSCs to have a snapshot source rather than a volume source, but if that's not necessary for triggering a deletion, then :+1:.

Related, is there any reason to sync in the VolumeSnapshots? It sounds like we won't be using them for restore, and directly deleting the VSCs should be sufficient, per the previous PR - right?

Reply:
I second the question about restoring the VolumeSnapshots - my understanding was that we'd just sync the VSCs in order to have a way to garbage collect storage-side snapshots associated with an expired backup.

Reply:
That's correct. We don't have to restore volumesnapshots. I'll remove that. Thanks for catching!